cpp/oneapi/dal/test/array.cpp

```diff
@@ -0,0 +1,332 @@
+        REQUIRE(arr[i] == Approx(0.0f));
+    }
+}
+
+TEST("can_construct_array_of_ones") {
+    auto arr = array<float>::full(5, 1.0f);
+
+    REQUIRE(arr.get_count() == 5);
+    REQUIRE(arr.has_mutable_data());
+
+    for (int32_t i = 0; i < arr.get_count(); i++) {
+        REQUIRE(arr[i] == Approx(1.0f));
+    }
+}
+
+TEST("can_construct_array_from_raw_pointer") {
+    constexpr int64_t size = 10;
+    auto ptr = new float[size];
+    for (int64_t i = 0; i < size; i++) {
+        ptr[i] = float(i);
+    }
+
+    array arr(ptr, size, [](auto ptr) {
+        delete[] ptr;
+    });
+
+    REQUIRE(arr.get_count() == size);
+    REQUIRE(arr.has_mutable_data());
+
+    for (int32_t i = 0; i < arr.get_count(); i++) {
+        REQUIRE(arr[i] == Approx(ptr[i]));
+    }
+}
+
+TEST("can_construct_array_reference") {
+    auto arr = array<float>::zeros(5);
+    array<float> arr2 = arr;
+
+    REQUIRE(arr.get_count() == arr2.get_count());
+    REQUIRE(arr.get_data() == arr2.get_data());
+    REQUIRE(arr.has_mutable_data() == arr2.has_mutable_data());
+    REQUIRE(arr.get_mutable_data() == arr2.get_mutable_data());
+
+    for (int32_t i = 0; i < arr.get_count(); i++) {
+        REQUIRE(arr[i] == Approx(arr2[i]));
+    }
+}
+
+TEST("can reset array") {
+    auto arr = array<float>::zeros(5);
+    arr.reset();
+
+    REQUIRE(arr.get_count() == 0);
+    REQUIRE(!arr.has_mutable_data());
+}
+
+TEST("can reset array with bigger size") {
+    auto arr = array<float>::zeros(5);
+    arr.reset(10);
+
+    REQUIRE(arr.get_count() == 10);
+    REQUIRE(arr.has_mutable_data());
+}
+
+TEST("can reset array with smaller size") {
+    auto arr = array<float>::zeros(5);
+    arr.reset(4);
+
+    REQUIRE(arr.get_count() == 4);
+    REQUIRE(arr.has_mutable_data());
+}
+
+TEST("can reset array with raw pointer") {
+    auto arr = array<float>::zeros(5);
+
+    constexpr int64_t size = 10;
+    auto ptr = new float[size];
+    arr.reset(ptr, size, [](auto ptr) {
+        delete[] ptr;
+    });
+
+    REQUIRE(arr.get_count() == size);
+    REQUIRE(arr.has_mutable_data());
+    REQUIRE(arr.get_mutable_data() == ptr);
+    REQUIRE(arr.get_data() == ptr);
+}
+
+TEST("can reset array with non owning raw pointer") {
+    auto arr = array<float>::zeros(5);
+
+    constexpr int64_t size = 10;
+    const float* ptr = new float[size];
+    arr.reset(array<float>(), ptr, size);
+
+    REQUIRE(arr.get_count() == size);
+    REQUIRE(arr.get_data() == ptr);
+    REQUIRE(!arr.has_mutable_data());
+    // ASSERT_THROW(arr.get_mutable_data(), std::bad_variant_access);
+
+    delete[] ptr;
+}
+
+TEST("can make owning array from non owning readonly") {
+    array<float> arr;
+
+    float data[] = { 1.f, 2.f, 3.f };
+    arr.reset(array<float>(), (const float*)(data), 3);
+
+    REQUIRE(arr.get_count() == 3);
+    REQUIRE(arr.get_data() == data);
+    REQUIRE(!arr.has_mutable_data());
+    // ASSERT_THROW(arr.get_mutable_data(), std::bad_variant_access);
```

Comment:
Uncomment or place todo here?
I think `bad_variant_access` is an implementation detail and we shall throw something like `dal::logic_error`